extractor/filesystem/os/apk/extractor.go

```diff
@@ -129,67 +126,91 @@ func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]
+			}
+
+			group[key] = val
+			continue
+		}
+
+		// check both that line is empty and we have filled out data in group
+		// this avoids double empty lines returning early
+		if line == "" && len(group) > 0 {
+			return group, nil
+		}
+	}
+
+	// Err() could only be non nil when Scan() returns false
+	return group, scanner.Err()
+}
+
 func (e Extractor) extractFromInput(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
 	m, err := osrelease.GetOSRelease(input.FS)
 	if err != nil {
 		log.Errorf("osrelease.ParseOsRelease(): %v", err)
 	}
 
-	rd := textproto.NewReader(bufio.NewReader(input.Reader))
-	pkgs := []*extractor.Inventory{}
+	scanner := bufio.NewScanner(input.Reader)
+	inventories := make([]*extractor.Inventory, 0)
+
 	for eof := false; !eof; {
-		// Return if canceled or exceeding deadline.
 		if err := ctx.Err(); err != nil {
-			return pkgs, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+			return nil, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
 		}
 
-		h, err := rd.ReadMIMEHeader()
+		record, err := parseSingleApkRecord(scanner)
 		if err != nil {
-			if errors.Is(err, io.EOF) {
-				// We might still have one more line of data
-				// so return only after it's been parsed.
-				eof = true
-			} else {
-				log.Errorf("Failed to extract all APK packages: %v", err)
-				return pkgs, nil
-			}
+			return nil, fmt.Errorf("error while parsing apk status file %q: %w", input.Path, err)
 		}
-		pkgName := h.Get("P")
-		version := h.Get("V")
-		if pkgName == "" || version == "" {
-			if !eof { // Expected when reaching the last line.
-				log.Warnf("APK package name or version is empty (name: %q, version: %q)", pkgName, version)
-			}
-			continue
-		}
-		originName := h.Get("o")
-		maintainer := h.Get("m")
-		arch := h.Get("A")
-		license := h.Get("L")
-		commit := h.Get("c")
-		var sourceCode *extractor.SourceCodeIdentifier
-		if commit != "" {
-			sourceCode = &extractor.SourceCodeIdentifier{
-				Commit: commit,
-			}
+
+		if len(record) == 0 {
+			break
 		}
-		pkgs = append(pkgs, &extractor.Inventory{
-			Name:    pkgName,
-			Version: version,
+
+		var pkg = &extractor.Inventory{
+			Name:    record["P"],
+			Version: record["V"],
 			Metadata: &Metadata{
-				PackageName:  pkgName,
-				OriginName:   originName,
 				OSID:         m["ID"],
 				OSVersionID:  m["VERSION_ID"],
-				Maintainer:   maintainer,
-				Architecture: arch,
-				License:      license,
+				PackageName:  record["P"],
+				OriginName:   record["o"],
+				Architecture: record["A"],
+				License:      record["L"],
+				Maintainer:   record["m"],
 			},
-			SourceCode: sourceCode,
-			Locations:  []string{input.Path},
-		})
+			Locations: []string{input.Path},
+		}
+
+		if commit, ok := record["c"]; ok {
+			pkg.SourceCode = &extractor.SourceCodeIdentifier{
+				Commit: commit,
+			}
+		}
+
+		if pkg.Name == "" || pkg.Version == "" {
```

Comment:
You could also do this check before creating the inventory, like diff base did. Any reason for not doing this?

Reply:
It's just slightly neater by avoiding making two additional local variables. I don't think performance wise it matters too much either way since this is an edge case?

Reply:
sg